File: Algorithm/Code_Materials/c_or_cpp/charpter10_sort/level3/merge_sort.cpp
#include <stdio.h>
#include <stdlib.h>

void merge(int left[], int leftLength, int right[], int rightLength, int result[]) {
    int ll = 0, rr = 0, resIdx = 0;
  
    while (ll < leftLength && rr < rightLength) {
        if (left[ll] < right[rr]) {
            result[resIdx++] = left[ll++];
        } else {
            result[resIdx++] = right[rr++];
        }
    }
  
    while (ll < leftLength) {
        result[resIdx++] = left[ll++];
    }
  
    while (rr < rightLength) {
        result[resIdx++] = right[rr++];
    }
}

void mergeSort(int arr[], int length) {
    if (length <= 1) {
        return;
    }
  
    int mid = length / 2;
  
    int left[mid];
    int right[length - mid];
  
    for (int i = 0; i < mid; i++) {
        left[i] = arr[i];
    }
  
    for (int i = mid; i < length; i++) {
        right[i - mid] = arr[i];
    }
  
    mergeSort(left, mid);
    mergeSort(right, length - mid);
  
    merge(left, mid, right, length - mid, arr);
}

void printArray(int arr[], int length) {
    for (int i = 0; i < length; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main() {
    int arr[] = {43, 23, 10, 7, 36, 87, 5, 19};
    int length = sizeof(arr) / sizeof(arr[0]);
  
    printf("Original array: ");
    printArray(arr, length);
  
    mergeSort(arr, length);
  
    printf("Sorted array: ");
    printArray(arr, length);
  
    return 0;
}

File: Algorithm/Leetcode/Code_Materials/python_source/charpter10_sort/level3/merge_sort.py
# coding=utf-8
def merge(left, right):
    ll, rr = 0, 0
    result = []
    while ll < len(left) and rr < len(right):
        if left[ll] < right[rr]:
            result.append(left[ll])
            ll += 1
        else:
            result.append(right[rr])
            rr += 1
    result+=left[ll:]
    result+=right[rr:]
    return result

def merge_sort(alist):
    if len(alist) <= 1:
        return alist
    num = len(alist) // 2   # 从中间划分两个子序列
    left = merge_sort(alist[:num]) # 对左侧子序列进行递归排序
    right = merge_sort(alist[num:]) # 对右侧子序列进行递归排序
    return merge(left, right) #归并

# 测试方法
if __name__ == '__main__':
    tesl = [1, 3, 45, 23, 23, 12, 43, 45, 33, 21]
    print(merge_sort(tesl))